src/transformers/tokenization_utils.py

```diff
@@ -37,6 +40,58 @@
 TOKENIZER_CONFIG_FILE = "tokenizer_config.json"
 
 
+@contextmanager
+def truncate_and_pad(
+    tokenizer: BaseTokenizer,
+    max_length: int,
+    stride: int,
+    strategy: str,
+    pad_to_max_length: bool,
+    padding_side: str,
+    pad_token_id: int,
+    pad_token_type_id: int,
+    pad_token: str,
+):
+    """
+    This contextmanager is in charge of defining the truncation and the padding strategies and then
+    restore the tokenizer settings afterwards.
+
+    :param tokenizer:
+    :param max_length:
+    :param stride:
+    :param strategy:
+    :param pad_to_max_length:
+    :param padding_side:
+    :param pad_token_id:
+    :param pad_token_type_id:
+    :param pad_token:
+    :return:
+    """
+
+    # Handle all the truncation and padding stuff
+    if max_length is not None:
+        tokenizer.enable_truncation(max_length, stride=stride, strategy=strategy)
+
+        if pad_to_max_length and pad_token:
+            tokenizer.enable_padding(
+                max_length=max_length,
```

Comment:
If we don't specify the `max_length` here, it will pad to the longest of a batch.

Reply:
Ok, so I slightly changed the code here.

__Before__ max_length set to `None` would disable both truncation and padding paths.

__Now__ max_length just controls the truncation strategy and __padding is not related anymore to the max_length value__. Providing `max_length = None` and `pad_to_max_length = True` will map to the behaviour you described. 

See https://github.com/huggingface/transformers/pull/2674/commits/17e1679e4c421fe70e40d631bf233bd6faa4abea#diff-66606105b55d0ec62ae34112ea3e3d20R75